guac-clearly-defined-certifier.md

```diff
@@ -0,0 +1,74 @@
+---
+
+### Overview
+
+GUAC (Graph for Understanding Artifact Composition) integrates with **ClearlyDefined** (https://clearlydefined.io/?sort=releaseDate&sortDesc=true) to enhance supply chain transparency by retrieving accurate license data for software dependencies. This functionality helps organizations make informed decisions about software licenses when managing their dependencies.
+
+### How GUAC Pulls from ClearlyDefined
+
+GUAC accesses ClearlyDefined data through a **certifier process** instead of a collector model. This choice ensures that queries are **re-run on a scheduled basis** to reflect any updated license information, given that new software releases can alter licensing data over time. The certifier’s role complements GUAC’s use of SBOM (Software Bill of Materials) data, allowing it to validate or augment license details.
+
+- **Invocation**: 
+  - The certifier runs periodically or can be triggered during **SBOM ingestion** to pull data directly from ClearlyDefined.
+  - However, including ClearlyDefined queries during SBOM ingestion may **slow down processing**, especially for larger datasets.
+  - For further details visit https://docs.guac.sh/guac-configuration/#certifier-configuration.
+
+- **Mapping and Query System**:  
+  - ClearlyDefined identifies software using **coordinates**, which GUAC maps to **pURLs (package URLs)**. A custom mapping library facilitates this conversion to align the two systems. 
+
+### What is and is not Supported
+
+1. **Supported Features**:
+   - GUAC retrieves detailed license data from ClearlyDefined.
+   - Both SBOM-derived and ClearlyDefined license data are preserved, allowing users to make manual decisions in case of conflicts.
+   - **CertifyLegal nodes** are created for both sources, ensuring complete graph observability.
+
+2. **Limitations**:
+   - **Batched queries** are not supported by ClearlyDefined’s API, which can result in rate limits and slower performance for large dependency graphs.
+   - The system does not automatically resolve conflicting data between SBOMs and ClearlyDefined; users must determine the most trustworthy source themselves.
+
+### Invocation Process
+
+GUAC integrates with ClearlyDefined in three ways:  
+
+1. **Scheduled Certifier Execution**  
+   - Automatically runs at specified intervals to keep the dependency data up to date.  
+
+2. **On-Demand during SBOM Ingestion**  
+   - Queries ClearlyDefined in real time during the ingestion of SBOMs, which may slow down the processing.  
+
+3. **Manual Command Line Execution**  
+   - Users can directly invoke the certifier using:  
+     ```bash
+     guacone certifier cd --input <path> --output <path> [options]
+     ```  
+   - This method provides greater control, allowing users to run queries whenever needed.  
+
+---
+
+Below is a table of the supported **command-line arguments** for the **ClearlyDefined certifier**. These flags allow users to control input, output, and execution behavior. 
```

Comment:
Where did you get this list of arguments, because it does not match the output of `guacone certifier cd --help`?

Reply:
could you please provide the ouputs of guacone certifier cd --help. 
I went through the blogs and code base of certifier but couldn't figure out a relevant change.

Reply:
```bash
bcotton@Bens-MacBook-Air ~/g/guac (main)> guacone certifier cd --help
runs the clearly defined certifier

Usage:
  guacone certifier cd [flags]

Flags:
      --certifier-batch-size int   sets the batch size for pagination query for the certifier (default 60000)
      --certifier-latency string   sets artificial latency on the certifier. Defaults to empty string (not enabled) but can set m, h, s...etc
  -h, --help                       help for cd
  -l, --last-scan int              hours since the last scan was run. If not set, run on all packages/sources (default 4)

Global Flags:
      --add-license-on-ingest   if enabled, the ingestor will query and ingest clearly defined for licenses. Warning: This will increase ingestion times
      --add-vuln-on-ingest      if enabled, the ingestor will query and ingest OSV for vulnerabilities. Warning: This will increase ingestion times
      --csub-addr string        address to connect to collect-sub service (default "localhost:2782")
      --csub-tls                enable tls connection to the server
      --csub-tls-skip-verify    skip verifying server certificate (for self-signed certificates for example)
      --gql-addr string         endpoint used to connect to graphQL server (default "http://localhost:8080/query")
      --header-file string      a text file containing HTTP headers to send to the GQL server, in RFC 822 format
  -i, --interval string         if polling set interval, m, h, s, etc. (default "5m")
  -p, --poll                    sets the collector or certifier to polling mode
```